Translate: useId (#661)
reactjs/ko.react.dev: src/content/reference/react/useId.md

  // ...
```

[아래에서 더 많은 예시를 확인하세요.](#usage)

#### 매개변수 {/*parameters*/}

`useId`는 어떤 매개변수도 받지 않습니다.

#### 반환값 {/*returns*/}

`useId`를 호출한 특정 컴포넌트와 특정 `useId`에 관련된 고유 ID 문자열을 반환합니다.

#### 주의 사항 {/*caveats*/}

* `useId`는 Hook이므로 **컴포넌트의 최상위** 또는 커스텀 Hook에서만 호출할 수 있습니다. 반복문이나 조건문에서는 사용할 수 없습니다. 필요한 경우 새로운 컴포넌트를 추출하고 해당 컴포넌트로 state를 이동해서 사용할 수 있습니다.

* `useId`를 리스트의 **key를 생성하기 위해 사용하면 안 됩니다**. [Key는 데이터에서 생성해야 합니다.](/learn/rendering-lists#where-to-get-your-key)

---

## 사용법 {/*usage*/}

<Pitfall>

> **`useId`를 리스트의 key를 생성하기 위해 사용하면 안 됩니다.** [Key는 데이터에서 생성해야 합니다.](/learn/rendering-lists#where-to-get-your-key)

Reviewer: ```suggestion
**`useId`를 리스트의 key를 생성하기 위해 사용하면 안 됩니다.** [Key는 데이터로 부터 생성해야 합니다.](/learn/rendering-lists#where-to-get-your-key)
```

</Sandpack>

[영상](https://www.youtube.com/watch?v=0dNzNcuEuOo)을 통해 보조 기술을 활용했을 때 사용자 경험의 차이점을 확인할 수 있습니다.

<Pitfall>

[서버 렌더링](/reference/react-dom/server)에서 **`useId`는 서버와 클라이언트에서 동일한 컴포넌트 트리가 필요합니다.** 서버와 클라이언트에서 렌더링하는 트리가 정확히 일치하지 않으면 생성된 ID는 일치하지 않습니다.

</Pitfall>

<DeepDive>

> #### useId가 카운터를 증가하는 것보다 나은 이유는 무엇일까요? {/*why-is-useid-better-than-an-incrementing-counter*/}

Reviewer: ```suggestion
#### useId의 카운터를 증가하는 것보다 나은 이유는 무엇일까요? {/*why-is-useid-better-than-an-incrementing-counter*/}
```

Author: @hg-pyun 님 리뷰 감사합니다 🙇‍♀️

다만 수정해 주신 문장에 대해 조금 염려되는 점이 있어서 이에 대해 여쭤보고 싶습니다.

해당 내용은 usdId Hook을 사용하는 것이 전역 변수인 카운터를 사용하는 것보다 더 나은 점을 설명하는 부분인데, "useId의 카운터를 증가하는 것보다 나은 이유는 무엇일까요?"라고 번역했을 때 (useId의 카운터)로 해석되어 useId Hook의 자체적인 카운터로 잘못 해석될까 봐 걱정되어 이 점에 대해 여쭤보고 싶습니다.

리뷰해 주신 거처럼 저도 해당 문장이 어색하다고 생각하는데 혹시 "useId를 사용하는 것이 카운터를 증가하는 것보다 나은 이유는 무엇일까요?"라고 번역하는 것은 어떨지 제안하고 싶습니다.

혹시라도 제가 염려한 것처럼 잘못 해석될 여지가 없다면 수정해 주신 내용으로 반영하겠습니다 :)

Author: 확인해 주셔서 감사합니다.

해당 문장 포함해서 리뷰해 주신 내용 모두 반영했습니다 :)

`useId`가 `nextId++`처럼 글로벌 변수를 증가하는 것보다 나은 이유에 대해 궁금할 수 있습니다.

`useId`의 주요 이점은 React가 [서버 렌더링](/reference/react-dom/server)과 함께 작동하도록 보장한다는 것입니다. 서버 렌더링을 하는 동안 컴포넌트는 HTML 결과물을 생성합니다. 이후, 클라이언트에서 [hydration](/reference/react-dom/client/hydrateRoot)이 HTML 결과물에 이벤트 핸들러를 연결합니다. hydration이 동작하려면 클라이언트의 출력이 서버 HTML과 일치해야 합니다.

클라이언트 컴포넌트의 hydrated 순서가 서버 HTML이 생성된 순서와 일치하지 않을 수 있기 때문에 카운터 증가로 이를 보장하기는 매우 어렵습니다. `useId`를 사용하면 hydration이 동작하고 서버와 클라이언트 간에 출력이 일치하는 것을 보장할 수 있습니다.

React에서 `useId`는 호출한 컴포넌트의 "부모 경로"에서 생성됩니다. 클라이언트와 서버 트리가 동일한 경우 렌더링 순서와 관계없이 "부모 경로"가 일치하는 이유입니다.

</DeepDive>

---

> ### 연관된 여러 개의 엘리먼트의 ID 생성하기 {/*generating-ids-for-several-related-elements*/}

Reviewer: ```suggestion
### 여러 개의 연관된 엘리먼트의 ID 생성하기 {/*generating-ids-for-several-related-elements*/}
```

> 연관된 여러 개의 엘리먼트에 ID를 전달하는 과정이 필요할 때 `useId`를 사용해서 공유 접두사를 생성할 수 있습니다.

Reviewer: ```suggestion 
여러 개의 연관된 엘리먼트에 ID를 전달하는 과정이 필요할 때 `useId`를 사용해서 공유 접두사를 생성할 수 있습니다.
```
